fix: stock depot selection issues

jniles opened the pull request:
This PR addresses multiple issues associated with depot selection in the stock management modules.  It implements the suggestion in #2003 and removes the `bhDepotDropdown`.

The basic idea is to make depot selection like cashbox selection, where a user is presented a modal if they have no depot selected. This modal is shown below:
![depotselectionfirsttime](https://user-images.githubusercontent.com/896472/30426885-9814932a-9945-11e7-8042-2bb51e68796e.png)
_Fig 1: Depot Selection if no depot is selected_

Depots are cached between selections, so a page will always use the selected depot.  They are not (yet) cached between pages ... although this would be as simple as changing the cache key.

Changing the depot is accessible via the dropdown menu as on most pages:
![depotselectionmenu](https://user-images.githubusercontent.com/896472/30426986-02265a5a-9946-11e7-927d-f4c2cf3b06be.png)
_Fig 2: Depot Selection Menu_

Any subsequent selections bring up the following modal:
![depotselectionallothertimes](https://user-images.githubusercontent.com/896472/30427033-1df33834-9946-11e7-8579-5b20c24671be.png)
_Fig 3: Depot Selection Modal All Other Times_

Closes #2003.  Closes #2094.  Closes #1986.

jniles committed:
feat(stock): implement depot selection modal

This commit implements a depot selection modal for the stock modules.
The depot is cached between sessions, but if there is no depot, a user
is immediately directed to the open modal to choose a depot.  This modal
cannot be dismissed if the user has not chosen a depot.

Additionally, a dropdown menu has been implemented on the stock page to
put the "Change Depot" switch in.

Ideally, this should also clear and reset the stock form.  However,
since that isn't even implemented for the "submit" button, this feature
will be added later.

The prototype has been implemented on the Stock Entry page.

jniles committed:
feat(stock): depot selection modal for exits

This commit implements the depot selection modal for stock exits.  It
follows the same principles of the stock entry.

jniles committed:
feat(stock): adjustment select depot modal

This commit implements the depot selection modal on the stock adjustment
page.  It fixes a bug where modals were not correctly dismissed because
of the $state transitions taking place.

jniles committed:
stock: remove bhDepotDropdown

This commit removes bhDepotDropdown which has been superceded by the new
depot selection modal.

Closes #1986.  Closes #2003.

jniles committed:
test(stock): e2e test for depot selection

This commit implements a shared page for selecting the depot.

jniles committed:
chore(stock): implement i18n translations

This commit implements i18n translations for the depot selection
modal.

jniles committed:
fix(stock): invalid tests locator

This commit replaces custom bhDepotSelect code with FU.uiSelect() to
ensure that the test correctly references the underlying element.

DedrickEnc commented on client/src/js/services/StockService.js:
```diff
@@ -45,40 +45,39 @@ function StockService(Api, Filters, AppCache, Periods, $httpParamSerializer, Lan
   assignMovementDefaultFilters();
 
-  // creating an an object of filter to avoid method duplication
+  // creating an object of filter to avoid method duplication
```
good catch

DedrickEnc commented on client/src/modules/cash/cash.html:
```diff
@@ -11,7 +11,7 @@
         <div uib-dropdown dropdown-append-to-body data-action="open-tools">
           <a class="btn btn-default" uib-dropdown-toggle>
-            <span translate>FORM.LABELS.MENU</span> <span class="caret"></span>
+            <span class="fa fa-bars"></span> <span class="hidden-xs" translate>FORM.LABELS.MENU</span> <span class="caret"></span>
```
A good enhancement

DedrickEnc commented on client/src/modules/depots/depots.service.js:
```diff
@@ -10,7 +10,28 @@ DepotService.$inject = ['PrototypeApiService'];
+   * @description
+   * Opens the selection modal to allow a user to select a depot.
+   *
+   * @returns Promise - a promise containing the depot.
```
or a promise containing a selected depot

jniles commented on client/src/i18n/en/depot.json:
```diff
@@ -1,22 +1,30 @@
+      "TITLE" : "Depot Management"
+    },
+    "NO_DEPOT" : "No Depot",
+    "NOT_ALLOWED_ACTIONS" : "This depot is protect.  You cannot modify its properties.",
```
Whoops, this should say "this depot is protected." instead of "this depot is protect."

jniles commented on client/src/i18n/en/depot.json:
```diff
@@ -1,22 +1,30 @@
+      "SELECTION" : "Depot Selection",
+      "NO_DEPOTS" : "Do not see any depots?",
+      "SELECT_A_DEPOT" : "Select a depot",
+      "NO_DEFAULT_DEPOT" : "Oops!  We need you to choose a depot.",
```
This would be better said ad "You need to choose a depot."

jniles commented on client/src/modules/stock/adjustment/adjustment.js:
```diff
@@ -127,12 +122,32 @@ function StockAdjustmentController(Depots, Inventory, Notify,
+    // make sure that the depot is loaded if it doesn't exist at startup.
+    if (vm.depot) {
+      setupStock();
+      loadInventories(vm.depot);
```
Technically, this repeated code could be combined into a single function:
```js
function setupStock(depot) {
  // old code
  loadInventories(vm.depot);
  checkValidity();
}

// later..

if (vm.depot) {
  setupStock(vm.depot);
} else {
  changeDepot()
    .then(setupStock);
}
```

Slight code reduction here by about two lines.

DedrickEnc replied:
hmmmm .... yes

jniles commented on client/src/modules/stock/depot-selection.modal.html:
```diff
@@ -0,0 +1,72 @@
+      <strong> <i class="fa fa-warning"></i> <span translate>DEPOT.MODAL.NO_DEFAULT_DEPOT</span></strong>
+      <br />
+      <span translate>DEPOT.MODAL.SELECTION_EXPLANATION</span>
+      <a href ui-sref="depots.create"><b><span translate>DEPOT.ADD_DEPOT</span></b></a>.
```
It would be nice if this was a colored yellow like the rest of the alert.  It is currently blue, which really stands out and clashes with the rest of the colors.

jniles commented on client/src/modules/stock/entry/entry.html:
```diff
@@ -95,8 +115,8 @@
-      ui-grid="StockCtrl.gridOptions" 
+    <div id="stock-entry-grid"
+      ui-grid="StockCtrl.gridOptions"
       style="height: 300px; width: 100%;"
```
It would be nice if these styles were contained in the `bhima-bootstrap.less` file.  This should probably be a separate PR that addresses it here and in all other modules that have this style of `ui-grid`.

jniles commented on test/end-to-end/stock/stock.lots.spec.js:
```diff
@@ -51,6 +51,7 @@ function StockLotsRegistryTests() {
   });
 
   it('find lots in depot principal', () => {
+    browser.debugger();
```
Whoops.  :dancer:

DedrickEnc replied:
????

DedrickEnc commented on client/src/modules/stock/adjustment/adjustment.html:
```diff
@@ -52,7 +72,7 @@
             name="description"
             ng-model="StockCtrl.movement.description"
-            placeholder="{{ ::'FORM.PLACEHOLDERS.DESCRIPTION' | translate }}"
+            placeholder="{{ 'FORM.PLACEHOLDERS.DESCRIPTION' | translate }}"
```
Why you remove it?

jniles replied:
`::` is a one time binding.  It would bind the untranslated text in about 30% of the page loads for me.

DedrickEnc commented on client/src/modules/stock/depot-selection.ctrl.js:
```diff
@@ -0,0 +1,59 @@
+  function selectDepot(uuid) {
+    var selected;
+
+    vm.depots.forEach(function (d) {
```
It seems like this can be performed by ```Array.filter``` like : 

```
var selected = vm.depots.filter(function (d) {
  return d.uuid === uuid;
}).pop();
```

Because we are sure to always match a depot with the provided ```uuid```

DedrickEnc commented on client/src/modules/stock/depot-selection.ctrl.js:
```diff
@@ -0,0 +1,59 @@
+  vm.hasSelectedDepot = hasSelectedDepot;
+  vm.loading = false;
+
+  // this is a one-time message shown to the user if they do not have a cached depot.
```
Where is the message? I think it is just a flag and the message is may be at the client side right?
If I am right, I think the comment should be updated.

jniles replied:
Ah, it is a boolean for showing the message.  The  message is actually in the HTML, hidden by an `ngIf` that listens on this variable.  The comment  could better describe this, I agree.

DedrickEnc commented on client/src/modules/stock/depot-selection.modal.html:
```diff
@@ -0,0 +1,72 @@
+  <label class="control-label" translate>DEPOT.MODAL.SELECT_A_DEPOT</label>
+
+  <!-- list of depots -->
+  <ul class="list-group">
```
What will happen if we have an institution with 10 depots ? the modal will still fit well in this situation? why not  to use select? (I like this user interface, I am just thinking in a laud voice).

jniles replied:
In theory, there shouldn't be _too_ many depots.  But with more depots, it looks like this:

![selectdepotwithatonofdepots](https://user-images.githubusercontent.com/896472/30484577-b34019ca-9a22-11e7-9b07-ef1de12383f5.png)
_Fig 1: Select depot with 14 depots_

DedrickEnc commented on client/src/modules/stock/entry/entry.html:
```diff
@@ -2,12 +2,32 @@
+        <span translate>STOCK.ENTRY</span>
+      </li>
+
+      <li class="title" ng-if="StockCtrl.depot.uuid">
```
Why not ```StockCtrl.selectedDepot```?

jniles replied:
I believe the previous version used `depot`, so I was attempting not to change the bindings too much :(.

DedrickEnc commented on client/src/modules/stock/entry/entry.js:
```diff
@@ -253,7 +261,32 @@ function StockEntryController(Depots, Inventory, Notify,
+      .catch(Notify.handleError);
+  }
+
+  /**
```
Good,  it will be better to precise what are ```data``` and ```uuid``` also in the comments as we have precision about function name, description and return statement

DedrickEnc commented on client/src/modules/stock/entry/entry.js:
```diff
@@ -3,45 +3,49 @@ angular.module('bhima.controllers')
   vm.itemIncrement = 1;
   vm.enterprise = Session.enterprise;
   vm.maxLength = util.maxLength;
+  vm.maxDate = new Date();
```
It is seems like ```vm.maxDate``` is never used

jniles replied:
You are right!  It isn't ever used, but it should be - see https://github.com/IMA-WorldHealth/bhima-2.X/pull/2097 which fixes this in at least one module.

DedrickEnc commented on client/src/modules/stock/exit/exit.js:
```diff
@@ -6,141 +6,124 @@ StockExitController.$inject = [
-  vm.enterprise = Session.enterprise;
   vm.maxLength = util.maxLength;
+  vm.enterprise = Session.enterprise;
+  vm.maxDate = new Date();
```
is it used somewhere?

DedrickEnc commented on client/src/modules/stock/exit/exit.js:
```diff
@@ -6,141 +6,124 @@ StockExitController.$inject = [
-    vm.gridApi = api;
-  }
-
   // exit type
```
This comment tells almost nothing, I know it was there before this P.R

jniles replied:
Ha! :+1:

DedrickEnc commented on test/end-to-end/stock/stock.shared.page.js:
```diff
@@ -0,0 +1,59 @@
+ * @description
+ * Makes sure that the modal is either already open or opens it.
+ *
+ * TODO(@jniles) This isn't great practice ... it probably should be re-examined
```
LOL

DedrickEnc reviewed (commented):
@jniles
I have made some minors comments, you can go a head fix them or clarify me and then the P.R should be ready to be merge in.

jniles committed:
fix(stock): misc cleanup

This commit cleans up and optimizes the code, fixing typos and artifacts
left behind from testing.

DedrickEnc reviewed (approved):
LGTM
